src/qiskit_qec/circuits/css_code.py

```diff
@@ -0,0 +1,465 @@
+
+            gauges = [[], []]
+            stabilizers = [[], []]
+            logicals = [[], []]
+
+            for (
+                raw_ops,
+                ops,
+            ) in zip([raw_gauges, raw_stabilizers, raw_logicals], [gauges, stabilizers, logicals]):
+                for op in raw_ops:
+                    op = str(op)
+                    for j, pauli in enumerate(["X", "Z"]):
+                        if (op.count(pauli) + op.count("I")) == self.code.n:
+                            ops[j].append([k for k, p in enumerate(op[::-1]) if p == pauli])
+                is_css = is_css and (len(ops[0]) + len(ops[1])) == len(raw_ops)
+
+            # extra stabilizers: the product of all others
+            for j in range(2):
+                combined = []
+                for stabilizer in stabilizers[j]:
+                    combined += stabilizer
+                stabilizers[j].append([])
+                for q in combined:
+                    if combined.count(q) % 2:
+                        stabilizers[j][-1].append(q)
+
+            if is_css:
+                self.x_gauges = gauges[0]
+                self.z_gauges = gauges[1]
+                self.x_stabilizers = stabilizers[0]
+                self.z_stabilizers = stabilizers[1]
+                self.logical_x = logicals[0]
+                self.logical_z = logicals[1]
+            else:
+                raise QiskitQECError("Code is not obviously CSS.")
+
+        else:
+            # otherwise assume it has the info
+            self.x_gauges = self.code.x_gauges
+            self.z_gauges = self.code.z_gauges
+            self.x_stabilizers = self.code.x_stabilizers
+            self.z_stabilizers = self.code.z_stabilizers
+            self.logical_x = self.code.logical_x
+            self.logical_z = self.code.logical_z
+
+    def _z_gauge_measurements(self, qc, t, state):
+        creg = ClassicalRegister(len(self.z_gauges), name="round_" + str(t) + "_z_bits")
+        qc.add_register(creg)
+        for g, z_gauge in enumerate(self.z_gauges):
+            for q in z_gauge:
+                qc.cx(qc.qregs[0][q], qc.qregs[1][g])
+            if state[-1] == "n" and self._phenom:
+                qc.append(self._meas_error, [qc.qregs[1][g]])
+            qc.measure(qc.qregs[1][g], creg[g])
+            qc.reset(qc.qregs[1][g])
+
+    def _x_gauge_measurements(self, qc, t, state):
+        creg = ClassicalRegister(len(self.x_gauges), name="round_" + str(t) + "_x_bits")
+        qc.add_register(creg)
+        for g, x_gauge in enumerate(self.x_gauges):
+            for q in x_gauge:
+                qc.h(qc.qregs[0][q])
+                qc.cx(qc.qregs[0][q], qc.qregs[2][g])
+                qc.h(qc.qregs[0][q])
+            if state[-1] == "n" and self._phenom:
+                qc.append(self._meas_error, [qc.qregs[2][g]])
+            qc.measure(qc.qregs[2][g], creg[g])
+            qc.reset(qc.qregs[2][g])
+
+    def string2nodes(self, string, **kwargs):
```

Comment:
string2nodes is a method that appears all over the place and should probably be in a separate set of methods outside this class.

Reply:
`string2nodes` has to be custom written for each variant of a code circuit.